File: clients/beebium-typescript-client/src/exceptions.ts
/**
 * Error hierarchy for the Beebium TypeScript client.
 */

/** Base error for all Beebium client errors. */
export class BeebiumError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "BeebiumError";
    }
}

/** Failed to establish or maintain a gRPC connection to the server. */
export class ConnectionError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "ConnectionError";
    }
}

/**
 * The server's wire protocol does not match this client's. Raised at connect
 * time when the server's protocol fingerprint differs from the client's
 * compiled-in fingerprint. Install a server and client built from the same
 * protocol.
 */
export class ProtocolMismatchError extends ConnectionError {
    constructor(message: string) {
        super(message);
        this.name = "ProtocolMismatchError";
    }
}

/** The emulator server failed to start. */
export class ServerStartupError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "ServerStartupError";
    }
}

/** Could not locate a running emulator server. */
export class ServerNotFoundError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "ServerNotFoundError";
    }
}

/** An error occurred during a debugger operation. */
export class DebuggerError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "DebuggerError";
    }
}

/** A breakpoint or watchpoint condition expression is invalid. */
export class InvalidConditionError extends DebuggerError {
    constructor(message: string) {
        super(message);
        this.name = "InvalidConditionError";
    }
}

/** An invalid or failed memory access. */
export class MemoryAccessError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "MemoryAccessError";
    }
}

/** An operation exceeded its deadline. */
export class TimeoutError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "TimeoutError";
    }
}

/** An error related to disc operations. */
export class DiscError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "DiscError";
    }
}

/** An error related to Econet operations. */
export class EconetError extends BeebiumError {
    constructor(message: string) {
        super(message);
        this.name = "EconetError";
    }
}

File: clients/beebium-python-client/src/beebium/client/exceptions.py
# Copyright 2026 Robert Smallshire <robert@smallshire.org.uk>
#
# This file is part of Beebium.
#
# Beebium is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version. Beebium is distributed in the hope that it will
# be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with Beebium.
# If not, see <https://www.gnu.org/licenses/>.

"""Exception hierarchy for the beebium client."""


class BeebiumError(Exception):
    """Base exception for all beebium client errors."""

    pass


class ConnectionError(BeebiumError):
    """Failed to connect to the beebium server."""

    pass


class ProtocolMismatchError(ConnectionError):
    """The server's wire protocol does not match this client's.

    Raised at connect time when the server's protocol fingerprint differs from
    the client's compiled-in fingerprint. Install a server and client built from
    the same protocol.
    """

    pass


class ServerStartupError(BeebiumError):
    """The beebium-server process failed to start."""

    pass


class ServerNotFoundError(BeebiumError):
    """The beebium-server executable could not be found."""

    pass


class DebuggerError(BeebiumError):
    """A debugger operation failed."""

    pass


class InvalidConditionError(DebuggerError):
    """A breakpoint or watchpoint condition expression is invalid."""

    pass


class MemoryAccessError(BeebiumError):
    """A memory read or write operation failed."""

    pass


class TimeoutError(BeebiumError):
    """An operation timed out."""

    pass


class ScreenExpectTimeout(TimeoutError):
    """Expected screen text did not appear within the timeout (see Beebium.expect)."""

    pass


class DiscError(BeebiumError):
    """A disc operation failed."""

    pass


class EconetError(BeebiumError):
    """An Econet operation failed."""

    pass


class SerialError(BeebiumError):
    """A serial port operation failed."""

    pass


class ExtensionError(BeebiumError):
    """A peripheral-extension client-adapter operation failed."""

    pass


class ExtensionNotLoadedError(ExtensionError):
    """The requested extension is not loaded on the connected server.

    Raised by the extension bridge (``bbc.extensions[...]`` / ``Adapter.attach``)
    when no loaded extension matches the requested name or adapter type.
    """

    pass


class ExtensionAmbiguousError(ExtensionError):
    """More than one loaded extension matches a name/type request.

    Raised by the extension/transport bridges when a type or name key matches
    several loaded instances of the same kind. Address one by its instance id
    (``bbc.extensions["<id>"]`` / ``bbc.transport["<id>"]``) to disambiguate.
    """

    pass


class ExtensionAdapterNotInstalledError(ExtensionError):
    """No client adapter is installed for a loaded extension.

    Raised when the server has loaded an extension but no ``beebium.ext``
    adapter is registered to drive it (e.g. its adapter distribution is not
    installed). The extension is still visible via ``bbc.extensions.loaded``.
    """

    pass
